tensorflow/contrib/layers/python/layers/pooling.py

```diff
@@ -0,0 +1,169 @@
+  w = math_ops.cast(array_ops.gather(inputs_shape, 2), dtypes.int32)
+
+  if mode == 'max':
+    pooling_op = math_ops.reduce_max
+  elif mode == 'avg':
+    pooling_op = math_ops.reduce_mean
+  else:
+    msg = "Mode must be either 'max' or 'avg'. Got '{0}'"
+    raise ValueError(msg.format(mode))
+
+  result = []
+  n = pool_dimension
+  for row in range(pool_dimension):
+    for col in range(pool_dimension):
+      # start_h = floor(row / n * h)
+      start_h = math_ops.cast(
+        math_ops.floor(math_ops.multiply(math_ops.divide(row, n), math_ops.cast(h, dtypes.float32))), dtypes.int32)
+      # end_h = ceil((row + 1) / n * h)
+      end_h = math_ops.cast(
+        math_ops.ceil(math_ops.multiply(math_ops.divide((row + 1), n), math_ops.cast(h, dtypes.float32))), dtypes.int32)
+      # start_w = floor(col / n * w)
+      start_w = math_ops.cast(
+        math_ops.floor(math_ops.multiply(math_ops.divide(col, n), math_ops.cast(w, dtypes.float32))), dtypes.int32)
+      # end_w = ceil((col + 1) / n * w)
+      end_w = math_ops.cast(
+        math_ops.ceil(math_ops.multiply(math_ops.divide((col + 1), n), math_ops.cast(w, dtypes.float32))), dtypes.int32)
+      pooling_region = inputs[:, start_h:end_h, start_w:end_w, :]
+      pool_result = pooling_op(pooling_region, axis=(1, 2))
+      result.append(pool_result)
+  return result
+
+
+def spatial_pyramid_pooling(inputs, dimensions=None,
+                            mode='max', implementation='spp'):
+  """
+    Spatial pyramid pooling (SPP) is a pooling strategy to result in an output of fixed size.
+    It will turn a 2D input of arbitrary size into an output of fixed dimension.
+    Hence, the convolutional part of a DNN can be connected to a dense part
+    with a fixed number of nodes even if the dimensions of the input
+    image are unknown.
+    The pooling is performed over :math:`l` pooling levels.
+    Each pooling level :math:`i` will create :math:`M_i` output features.
+    :math:`M_i` is given by :math:`n_i * n_i`, with :math:`n_i` as the number
+    of pooling operations per dimension level :math:`i`.
+    The length of the parameter dimensions is the level of the spatial pyramid.
+
+  Args:
+    inputs: The tensor over which to pool. Must have rank 4.
+    dimensions: The list of bin sizes over which pooling is to be done.
+    mode: Pooling mode 'max' or 'avg'.
+    implementation: The implementation to use, either 'spp' or 'spp_optimized'.
+      `spp` is the original implementation from the paper, and supports variable
+      sizes of input vectors, which `spp_optimized` does not support.
+
+  Returns:
+    Output tensor.
+  """
+  layer = SpatialPyramidPooling(dimensions=dimensions,
+                                mode=mode,
+                                implementation=implementation)
+  return layer.apply(inputs)
+
+
+class SpatialPyramidPooling(base.Layer):
+  """
+    Spatial pyramid pooling (SPP) is a pooling strategy to result in an output of fixed size.
```

Comment:
Invalid docstring format (should start with a one-line summary, not a line break)